docs/notebooks/multi_objective_ehvi.pct.py

```diff
@@ -0,0 +1,201 @@
+            gpr = gpflow.models.GPR((single_obj_data.query_points, single_obj_data.observations), kernel, noise_variance=1e-5)
+            gpflow.utilities.set_trainable(gpr.likelihood, False)
+            gprs.append((create_model({
+            "model": gpr,
+            "optimizer": gpflow.optimizers.Scipy(),
+            "optimizer_args": {
+            "minimize_args": {"options": dict(maxiter=100)}}}), 1))
+
+        return ModelStack(*gprs)
+
+
+# %%
+models = {OBJECTIVE: build_stacked_independent_objectives_model(initial_data[OBJECTIVE], num_objective)}
+
+# %% [markdown]
+# ## Define the acquisition function
+# Here we utilize the [EHVI](https://link.springer.com/article/10.1007/s10898-019-00798-7): `ExpectedHypervolumeImprovement` acquisition function:
+
+# %%
+ehvi = ExpectedHypervolumeImprovement()
+rule: EfficientGlobalOptimization[Box] = EfficientGlobalOptimization(builder=ehvi.using(OBJECTIVE))
+
+# %% [markdown]
+# ## Run the optimization loop
+#
+# We can now run the optimization loop
+
+# %%
+num_steps = 30
+bo = trieste.bayesian_optimizer.BayesianOptimizer(observer, search_space)
+result = bo.optimize(num_steps, initial_data, models, acquisition_rule=rule)
+
+# %% [markdown]
+# To conclude, we visualize the queried data across the design space.  
+# The cross represented dots are initial sampled points by `search_space.sample`, the dots are bo-obtained points.
+
+# %%
+datasets = result.try_get_final_datasets()
+data_query_points = datasets[OBJECTIVE].query_points
+data_observations = datasets[OBJECTIVE].observations
+
+_, ax = plot_function_2d(
+    vlmop2,
+    mins,
+    maxs,
+    grid_density=100,
+    contour=True,
+    figsize=(12, 6),
+    title=["Obj 1", "Obj 2"],
+    xlabel="$X_1$",
+    ylabel="$X_2$",
+    colorbar=True,
+)
+plot_bo_points(data_query_points, ax=ax[0, 0], num_init=num_initial_points)
+plot_bo_points(data_query_points, ax=ax[0, 1], num_init=num_initial_points)
+plt.show()
+
+# %% [markdown]
+# Visualize in objective space. Purple dots denote the nondominated points.
+
+# %%
+plot_mobo_points_in_obj_space(data_observations, num_init=num_initial_points)
+plt.show()
+
+# %% [markdown]
+# We can also visualize how a performance metric evolved with respect to the number of BO iterations.
+# First, we need to define a performance metric. Many metrics have been considered for multi-objective optimization. Here, we use the log hypervolume difference, defined as the difference between the hypervolume of the true Pareto front and the hypervolume of the approximate Pareto front based on the bo-obtained data.
+
+# %% [markdown]
+#
+# $$
+# log_{10}\ \text{HV}_{diff} = log_{10}(\text{HV}_{\text{true}} - \text{HV}_{\text{bo-obtained}})
+# $$
+#
+
+# %% [markdown]
+# First we need to calculate the $\text{HV}_{\text{ideal}}$, this is approximated by calculating the hypervolume based on ideal pareto front:
+
+# %%
+ideal_pf = VLMOP2().gen_pareto_optimal_points(100)  # gen 100 pf points
```

Comment:
maybe detail a little bit what this does? how does it generate 100 points? are those random? it is generally not possible to do so unless the Pareto front or set is analytical (as it is here), so it might be worth saying it.

Reply:
Many thanks for the suggestions!   
Actually for VLMOP2 it is a bit special as CMIIW, the PF is defined through input space (ref p.50 of  the [paper](https://ieeexplore.ieee.org/document/501646)), so currently its Pareto front gen method is deterministic through `tf.linspace` in the input space, which seems not that consistent with DTLZ functions where the PF points are random generated. Perhaps could switch to uniform sample instead of `linspace`  if it makes more sense, while am not totally sure on this point.